arangod/RocksDBEngine/RocksDBEdgeIndex.cpp

```diff
@@ -282,57 +334,65 @@ class RocksDBEdgeIndexLookupIterator final : public IndexIterator {
       _lastKey = _keysIterator.value();
       TRI_ASSERT(_lastKey.isString());
       std::string_view fromTo = _lastKey.stringView();
+      std::string_view cacheKey;
 
       bool needRocksLookup = true;
       if (_cache) {
-        // Try to read from cache
-        auto finding =
-            _cache->find(fromTo.data(), static_cast<uint32_t>(fromTo.size()));
-        if (finding.found()) {
-          incrCacheHits();
-          needRocksLookup = false;
-          // We got sth. in the cache
-          VPackSlice cachedData(finding.value()->value());
-          TRI_ASSERT(cachedData.isArray());
-          if (cachedData.length() / 2 < limit) {
-            // Directly return it, no need to copy
-            _builderIterator = VPackArrayIterator(cachedData);
-            while (_builderIterator.valid()) {
-              TRI_ASSERT(_builderIterator.value().isNumber());
-              LocalDocumentId docId{
-                  _builderIterator.value().getNumericValue<uint64_t>()};
-
-              _builderIterator.next();
-
-              TRI_ASSERT(_builderIterator.valid());
-              TRI_ASSERT(_builderIterator.value().isString());
-              std::forward<F>(cb)(docId, _builderIterator.value());
-
-              _builderIterator.next();
-              limit--;
+        // try to read from cache
+
+        // build actual cache lookup key. this is either the value of fromTo,
+        // or a suffix of it in case fromTo starts with the cached collection
+        // name, and an empty string_view if the lookup value is invalid.
+        cacheKey = _index->buildCompressedCacheKey(cacheKeyCollection, fromTo);
+        TRI_ASSERT(cacheKey.empty() || cacheKey == fromTo ||
+                   fromTo.ends_with(cacheKey));
+
+        if (!cacheKey.empty()) {
+          // only look up a key in the cache if the key is syntactially valid.
+          // this is important because we can only tell syntactically valid keys
+          // apart from prefix-compressed keys.
+          auto finding = _cache->find(cacheKey.data(),
+                                      static_cast<uint32_t>(cacheKey.size()));
+
+          if (finding.found()) {
+            incrCacheHits();
+            needRocksLookup = false;
+            // We got sth. in the cache
+            VPackSlice cachedData(finding.value()->value());
+            TRI_ASSERT(cachedData.isArray());
+            if (cachedData.length() / 2 < limit) {
+              // Directly return it, no need to copy
+              _builderIterator = VPackArrayIterator(cachedData);
+              while (_builderIterator.valid()) {
+                handleSingleResult();
+                _builderIterator.next();
+                TRI_ASSERT(limit > 0);
```

Comment:
Could you explain this assert for me, please? Why is limit == 0 not ok here?

Reply:
Because we are subtracting 1 from it. If `limit == 0`, then it would underflow. That is exactly what the assertion should uncover if it happens.

Reply:
Overflow is defined for unsigned integer.

Anyway my question more about why if I have limit like 1000
Is not possible to produce exactly 1000 documents.

Oh I probably understand, it's because `< limit`, so our limit at least here is not included? `[0; limit)`

Reply:
We don't want the limit value to underflow, even if the underflow is defined here.
That is the motivation for the assert.

The entire `while` loop in addition is only executed if the condition `if (cachedIterator.size() / 2 < limit)` is true. That's why the limit will not go to 0 before the subtraction is happening.